Frameworks/CoreGraphics/DWriteFontBinaryDataCollectionLoader.mm

```diff
@@ -0,0 +1,170 @@
+
+using namespace Microsoft::WRL;
+
+/**
+ * Private helpers to append a CFErrorRef or nullptr to the end of a CFMutableArray if it exists
+ */
+static inline void __AppendErrorIfExists(CFMutableArrayRef errors, CFIndex errorCode) {
+    if (errors) {
+        woc::unique_cf<CFErrorRef> error{ CFErrorCreate(nullptr, kCFErrorDomainCocoa, errorCode, nullptr) };
+        CFArrayAppendValue(errors, error.get());
+    }
+}
+
+static inline void __AppendNullptrIfExists(CFMutableArrayRef errors) {
+    if (errors) {
+        CFArrayAppendValue(errors, nullptr);
+    }
+}
+
+/**
+ * Private class.
+ * Custom implementation of IDWriteFontFileEnumerator, which enumerates over internal data and returns a font file
+ */
+class DWriteFontFileEnumerator : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontFileEnumerator> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontFileEnumerator() {
+    }
+
+    HRESULT RuntimeClassInitialize(CFArrayRef fontDatas, std::vector<ComPtr<IDWriteFontFile>>* previouslyCreatedFiles) {
+        CFRetain(fontDatas);
+        m_fontDatas.reset(fontDatas);
+        m_previouslyCreatedFiles = previouslyCreatedFiles;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetCurrentFontFile(_Out_ IDWriteFontFile** fontFile) {
+        RETURN_HR_IF_NULL(E_POINTER, fontFile);
+        RETURN_HR_IF(E_ILLEGAL_METHOD_CALL, m_location < 0 || m_location > CFArrayGetCount(m_fontDatas.get()));
+
+        if (0 <= m_location && m_location < m_previouslyCreatedFiles->size()) {
+            m_previouslyCreatedFiles->at(m_location).CopyTo(fontFile);
+        } else {
+            ComPtr<IDWriteFactory> dwriteFactory;
+            RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+            ComPtr<DWriteFontBinaryDataLoader> loader;
+            RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataLoader>(&loader,
+                                                                           static_cast<CFDataRef>(
+                                                                               CFArrayGetValueAtIndex(m_fontDatas.get(), m_location))));
+            RETURN_IF_FAILED(dwriteFactory->RegisterFontFileLoader(loader.Get()));
+
+            int unused;
+            RETURN_IF_FAILED(dwriteFactory->CreateCustomFontFileReference(&unused, sizeof(unused), loader.Get(), fontFile));
+            m_previouslyCreatedFiles->emplace_back(*fontFile);
+        }
+
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE MoveNext(_Out_ BOOL* hasCurrentFile) {
+        RETURN_HR_IF_NULL(E_POINTER, hasCurrentFile);
+        *hasCurrentFile = ++m_location < CFArrayGetCount(m_fontDatas.get());
+        return S_OK;
+    }
+
+private:
+    int m_location = -1;
+
+    woc::unique_cf<CFArrayRef> m_fontDatas;
+    std::vector<ComPtr<IDWriteFontFile>>* m_previouslyCreatedFiles;
```

Comment:
The fact that this class, DWriteFontFileEnumerator, is a COM class makes this a dubious assumption. COM describes a shared ownership of objects. Presumably this guy is being handed out somewhere and the caller is free to hang on to it for as long as he/she wants. This could be longer than DWriteFontBinaryDataCollectionLoader is held onto. And in fact there is nothing else linking the Enumerator to the Loader which raises serious concerns about the assumptions you made in that comment. Can you explain more about how you are ensuring the validity of this pointer?

---
In reply to: [90908183](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90908183) [](ancestors = 90908183)